src/main/java/org/junitpioneer/jupiter/report/PioneerReportListener.java

```diff
@@ -0,0 +1,213 @@
+			if (propertyCache.containsKey(id)) {
+				propertyCache.get(id).add(property);
+			} else {
+				List<Property> currentProperties = new ArrayList<>();
+				currentProperties.add(property);
+				propertyCache.put(id, currentProperties);
+			}
+		}
+	}
+
+	@Override
+	public void executionFinished(TestIdentifier testIdentifier, TestExecutionResult testExecutionResult) {
+		if (testIdentifier.isTest()) {
+			// Store result in cache
+			testStatusCache.put(testIdentifier.getUniqueId(), testExecutionResult.getStatus().toString());
+		}
+	}
+
+	@Override
+	public void testPlanExecutionFinished(TestPlan testPlan) {
+		// Store plan to access it from methods
+		this.storedTestPlan = testPlan;
+		pioneerReport = new PioneerReport();
+
+		Set<TestIdentifier> allRoots = testPlan.getRoots();
+
+		for (TestIdentifier identifier : allRoots) {
+			if (identifier.isContainer()) {
+				pioneerReport.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				pioneerReport.getTestcase().add(processTest(identifier));
+			}
+		}
+		writeReport(pioneerReport);
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcontainer} for the report.
+	 *
+	 * @param container Object to be converted into {@code Testcontainer}
+	 * @return Created object
+	 */
+	Testcontainer processContainer(TestIdentifier container) {
+		Testcontainer testcontainer = new Testcontainer(container.getDisplayName());
+
+		// Add properties, if there are any
+		String uniqueId = container.getUniqueId();
+		Properties allProps = getProperties(uniqueId);
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcontainer.properties = allProps;
+		}
+
+		Set<TestIdentifier> allChildren = storedTestPlan.getChildren(uniqueId);
+
+		for (TestIdentifier identifier : allChildren) {
+			if (identifier.isContainer()) {
+				testcontainer.getTestcontainer().add(processContainer(identifier));
+			} else if (identifier.isTest()) {
+				testcontainer.getTestcase().add(processTest(identifier));
+			}
+		}
+
+		return testcontainer;
+	}
+
+	/**
+	 * Converts an {@code TestIdentifier} object into a {@code Testcase} for the report.
+	 *
+	 * @param test Object to be converted into {@code Testcase} element
+	 * @return Created object
+	 */
+	Testcase processTest(TestIdentifier test) {
+		String status = testStatusCache.getOrDefault(test.getUniqueId(), "UNKNOWN");
+		Testcase testcase = new Testcase(test.getDisplayName(), status);
+
+		// Add properties, if there are any
+		Properties allProps = getProperties(test.getUniqueId());
+
+		if (!allProps.getProperty().isEmpty()) {
+			testcase.setProperties(allProps);
+		}
+
+		return testcase;
+	}
+
+	/**
+	 * Retrieves all cached properties for an {@code TestIdentifier} by its unique id.
+	 *
+	 * @param testIdentifierUniqueId Unique id the properties should be loaded for.
+	 * @return Wrapper which contains all cached properties or an empty list.
+	 */
+	Properties getProperties(String testIdentifierUniqueId) {
```

Comment:
Maybe this could be an `Optional<Properties>` to indicate that you might _not_ always get back `Properties`?
E.g.:
```
Optional<Properties> getProperties(String id) {
    if (!propertyCache.containsKey(id)) {
        return Optional.empty();
    }
    Properties allProps = new Properties();
    allProps.getPropertyList().addAll(propertyCache.get(id));
    return Optional.of(allProps);
}
```

Reply:
I understand what you mean, but I'm not sure as I understand `Optional` is to be used to avoid NULL. And I never return NULL here.